be/src/service/service_be/lake_service.cpp

```diff
@@ -937,6 +938,124 @@ void LakeServiceImpl::compact(::google::protobuf::RpcController* controller, con
+        }
+    }
+
+    void wait() {
+        if (latch) {
+            latch->wait();
+        }
+    }
+
+    void count_down() {
+        if (latch) {
+            latch->count_down();
+        }
+    }
+
+    void write_combined_txn_log() {
+        if (final_status.ok()) {
+            final_status = starrocks::write_combined_txn_log(combined_txn_log);
+        }
+    }
+};
+
+static void aggregate_compact_cb(brpc::Controller* cntl, CompactResponse* response,
+                                 AggregateCompactContext* ac_context) {
+    // 1. release context
+    std::unique_ptr<brpc::Controller> cntl_guard(cntl);
+    std::unique_ptr<CompactResponse> response_guard(response);
+
+    DeferOp defer([&]() { ac_context->count_down(); });
+
+    // 2. check status
+    if (cntl->Failed()) {
+        ac_context->handle_failure("link rpc channel failed");
+        return;
+    } else {
+        if (response->status().status_code() != 0) {
+            ac_context->handle_failure("call compact failed");
+            return;
+        }
+    }
+
+    // 3. collect txn logs
+    ac_context->collect_txnlogs(response);
+}
+
+void LakeServiceImpl::aggregate_compact(::google::protobuf::RpcController* controller,
+                                        const ::starrocks::AggregateCompactRequest* request,
+                                        ::starrocks::CompactResponse* response, ::google::protobuf::Closure* done) {
+    brpc::ClosureGuard guard(done);
+    auto cntl = static_cast<brpc::Controller*>(controller);
+
+    if (request->requests_size() == 0) {
+        cntl->SetFailed("empty requests");
+        return;
+    }
+    if (request->compute_nodes_size() != request->requests_size()) {
+        cntl->SetFailed("compute nodes size not equal to requests size");
+        return;
+    }
+
+    AggregateCompactContext ac_context;
+    ac_context.latch = std::make_unique<BThreadCountDownLatch>(request->requests_size());
+
+    for (int i = 0; i < request->requests_size(); i++) {
+        const auto& single_req = request->requests(i);
+        const auto& compute_node = request->compute_nodes(i);
+        if (!compute_node.has_host() || !compute_node.has_brpc_port()) {
+            ac_context.handle_failure("compute node missing host/port");
+            ac_context.count_down();
+            continue;
+        }
+        brpc::Controller* node_cntl = new brpc::Controller();
+        CompactResponse* node_resp = new CompactResponse();
+        node_cntl->set_timeout_ms(single_req.timeout_ms());
+        butil::EndPoint endpoint;
+        std::string brpc_url = fmt::format("{}:{}", compute_node.host(), compute_node.brpc_port());
+        if (str2endpoint(brpc_url.c_str(), &endpoint)) {
+            ac_context.handle_failure("unknown endpoint, host=" + compute_node.host());
+            ac_context.count_down();
+            continue;
+        }
+        std::unique_ptr<brpc::Channel> channel(new brpc::Channel());
```

Comment:
should this go with the same channel as other brpc requests to other backend nodes?

Reply:
I don't get your point very well, but I think each channel will have its endpoint, so we can't use same channel for other BE?